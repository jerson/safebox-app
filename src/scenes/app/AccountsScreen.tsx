import React, { useState, useEffect } from 'react';
import {
  StyleSheet,
  ViewStyle,
  FlatList,
  StatusBar,
  RefreshControl
} from 'react-native';
import Colors from '../../modules/constants/Colors';
import Container from '../../components/ui/Container';
import Content from '../../components/ui/Content';
import Size from '../../modules/dimensions/Size';
import { AccountSingle } from '../../proto/services_pb';

import useAnimatedState from '../../components/hooks/useAnimatedState';
import HeaderIcon from '../../components/navigation/HeaderIcon';
import Client from '../../services/Client';
import Strings from '../../modules/format/Strings';
import Loading from '../../components/ui/Loading';
import AlertMessage from '../../components/ui/AlertMessage';
import EmptyAccounts from '../../components/help/EmptyAccounts';
import useFocusedScreen from '../../components/hooks/useFocusedScreen';
import AccountItem from '../../components/account/AccountItem';
import Timeout from '../../components/session/Timeout';
import { useNavigation } from 'react-navigation-hooks';

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.grey1
  } as ViewStyle,
  flatList: {
    flex: 1
  } as ViewStyle,
  alertMessage: {
    marginHorizontal: 20,
    marginBottom: 20
  } as ViewStyle,
  timeout: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0
  } as ViewStyle
});

function AccountsScreen() {
  const navigation = useNavigation();
  const [isLoading, setIsLoading] = useState(true);
  const [refreshing, setRefreshing] = useState(false);
  const [error, setError] = useAnimatedState('');
  const [accounts, setAccounts] = useAnimatedState<AccountSingle[]>([]);

  const [focused] = useFocusedScreen(navigation);

  const onRefresh = async () => {
    setRefreshing(true);
    await load();
    setRefreshing(false);
  };
  const load = async () => {
    try {
      const response = await Client.getAccounts();
      setAccounts(response.getAccountsList());
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
    setIsLoading(false);
  };

  useEffect(() => {
    focused && load();
  }, [focused]);

  return (
    <Container style={styles.container}>
      <StatusBar
        animated
        barStyle={'dark-content'}
        backgroundColor={Colors.grey2}
      />

      <Timeout style={styles.timeout} />
      {isLoading && <Loading margin={'large'} />}
      {!isLoading && (
        <FlatList
          ListHeaderComponent={() => {
            if (isLoading || !error) {
              return null;
            }

            return (
              <AlertMessage
                onTimeout={() => {
                  setError('');
                }}
                style={styles.alertMessage}
                message={error}
              />
            );
          }}
          extraData={{ isLoading, error }}
          refreshControl={
            <RefreshControl
              colors={[Colors.primary]}
              tintColor={Colors.primary}
              onRefresh={onRefresh}
              refreshing={refreshing}
            />
          }
          style={styles.flatList}
          contentContainerStyle={{
            minHeight: Size.getVisibleTabScreenHeight(),
            paddingTop: 80
          }}
          renderItem={({ item }) => {
            return <AccountItem item={item} />;
          }}
          data={accounts}
          keyExtractor={item => {
            return item.getId().toString();
          }}
          ListEmptyComponent={() => {
            return (
              <Content center>
                <EmptyAccounts />
              </Content>
            );
          }}
        />
      )}
    </Container>
  );
}

AccountsScreen.navigationOptions = ({ navigation }: any) => ({
  title: 'SafeBox',
  headerLeft: null,
  headerRight: (
    <HeaderIcon
      name={'plus-circle'}
      style={{ color: Colors.primaryLight }}
      onPress={() => {
        navigation.navigate('AddAccount');
      }}
    />
  )
});

export default AccountsScreen;
