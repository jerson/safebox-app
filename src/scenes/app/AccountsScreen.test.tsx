import React from 'react';
import * as renderer from 'react-test-renderer';
import AccountsScreen from './AccountsScreen';

const props: AccountsScreen = {};
test('renders without crashing', () => {
  const snapshot = renderer.create(<AccountsScreen {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<AccountsScreen {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
