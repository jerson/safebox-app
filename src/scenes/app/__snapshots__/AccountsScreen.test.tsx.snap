// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views 1`] = `
<_class
  style={
    Array [
      Object {
        "backgroundColor": "#fafdff",
        "flex": 1,
      },
      Object {
        "backgroundColor": "#fafdff",
      },
    ]
  }
>
  <BVLinearGradient
    colors={
      Array [
        4294639103,
        4294639103,
        4294639103,
        16449023,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      Array [
        Object {
          "justifyContent": "center",
          "padding": 20,
          "paddingTop": 5,
          "zIndex": 100,
        },
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      ]
    }
  >
    <_class
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 30,
          "marginBottom": 4,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#65748b",
              "flex": 1,
              "fontSize": 13,
            },
          ]
        }
      >
        Session expires in 
        00
      </Text>
      <_class
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#6781a6",
            "borderColor": "#6781a6",
            "borderRadius": 10,
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 0,
            "minWidth": 100,
            "opacity": 1,
            "padding": 4,
            "paddingHorizontal": 12,
            "paddingTop": 1,
            "zIndex": 1,
          }
        }
        testID="button_touchable"
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Nunito",
                "fontSize": undefined,
                "fontStyle": "normal",
                "fontWeight": "400",
              },
              Object {},
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 14,
                  "textAlign": "center",
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          Extend
        </Text>
      </_class>
    </_class>
    <_class
      style={
        Object {
          "backgroundColor": "#c2d0e7",
          "borderRadius": 6,
          "height": 6,
          "overflow": "hidden",
          "width": "100%",
        }
      }
    >
      <_class
        style={
          Object {
            "backgroundColor": "rgba(187, 202, 228, 1)",
            "borderRadius": 6,
            "height": 6,
            "width": "100%",
          }
        }
      />
    </_class>
  </BVLinearGradient>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListHeaderComponent={[Function]}
    contentContainerStyle={
      Object {
        "minHeight": 1220,
        "paddingTop": 80,
      }
    }
    data={Array []}
    disableVirtualization={false}
    extraData={
      Object {
        "error": "",
        "isLoading": false,
      }
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        colors={
          Array [
            "#4531cd",
          ]
        }
        onRefresh={[Function]}
        refreshing={false}
        tintColor="#4531cd"
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <_class>
      <_class
        onLayout={[Function]}
      />
      <_class
        style={
          Array [
            Object {
              "paddingHorizontal": 20,
            },
            Object {
              "justifyContent": "center",
            },
            Object {
              "flex": 1,
              "paddingTop": 20,
            },
            undefined,
          ]
        }
      >
        <_class
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "padding": 30,
            }
          }
        >
          <_class
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "#38414e",
                "borderColor": "#c2d0e7",
                "borderRadius": 22,
                "borderWidth": 1,
                "marginBottom": 10,
                "overflow": "hidden",
              }
            }
          >
            <Image
              resizeMode="stretch"
              source={
                Object {
                  "testUri": "../../../src/assets/images/logo.png",
                }
              }
              style={
                Object {
                  "height": 90,
                  "width": 90,
                }
              }
            />
          </_class>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Nunito",
                  "fontSize": undefined,
                  "fontStyle": "normal",
                  "fontWeight": "300",
                },
                Object {},
                Object {
                  "color": "#65748b",
                  "textAlign": "center",
                },
              ]
            }
          >
            The first thing you should do is add a 
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "Nunito",
                    "fontSize": undefined,
                    "fontStyle": "normal",
                    "fontWeight": "600",
                  },
                  Object {},
                  undefined,
                ]
              }
            >
              Secret account
            </Text>
          </Text>
          <_class
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#9066f5",
                "borderColor": "#9066f5",
                "borderRadius": 10,
                "flexDirection": "row",
                "justifyContent": "center",
                "margin": 0,
                "marginTop": 20,
                "opacity": 1,
                "padding": 12,
                "zIndex": 1,
              }
            }
            testID="button_touchable"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 12,
                  },
                  Array [
                    Object {
                      "color": "#FFFFFF",
                      "fontSize": 14,
                      "marginRight": 4,
                      "top": 3,
                    },
                    undefined,
                  ],
                  Object {
                    "fontFamily": "Feather",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              ï†µ
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "Nunito",
                    "fontSize": undefined,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                  },
                  Object {},
                  Array [
                    Object {
                      "color": "#FFFFFF",
                      "fontSize": 14,
                      "textAlign": "center",
                    },
                    undefined,
                    undefined,
                  ],
                ]
              }
            >
              Add first secret account
            </Text>
          </_class>
        </_class>
      </_class>
    </_class>
  </RCTScrollView>
</_class>
`;
